add support for climate action

marcelveldt opened the pull request:
Following hass implementation of climate, action represents the current action the climate device is perfoming, e.g. cooling or heating

## Description:
Current HVAC action/state changed to public attribute so it can be used in hass as hvac_action property on entity object.

**Related issue (if applicable):** 
fixes https://github.com/esphome/feature-requests/issues/392

## Checklist:
  - [X ] The code change is tested and works locally.

marcelveldt committed:
add support for climate action:

Following hass implementation of climate, action represents the current action the climate device is perfoming, e.g. cooling or heating

fix bang_bang climate:

make sure that the thresholds are always respected.
fixes the issue where the component would just keep on heating, regardless of the temperature range

marcelveldt commented:
The second commit fixes a bug in the bang_bang climate implementation. If the mode is not auto (so heating or cooling) it just ignores the boundaries setup so it just keeps on cooling or heating, regardless of the temperature ranges. This commit fixes that issue.

marcelveldt commented:
squashed commits for a cleaner PR. Changes are working on my system, together with the PR I've submitted for the python api.

glmnet commented on esphome/components/api/api.proto:
```diff
@@ -675,6 +675,7 @@ message ClimateStateResponse {
   float target_temperature_low = 5;
   float target_temperature_high = 6;
   bool away = 7;
+  ClimateMode action = 8;
```
This should be another enum for HA
https://github.com/home-assistant/home-assistant/blob/3efdf29dfa178838002201eaea258db78e6fa2ba/homeassistant/components/climate/const.py#L17

marcelveldt replied:
Yeah, probably it should but the previous implementation was also reusing the ClimateMode enum to store the current state. Do you want me to add another enum ?

glmnet replied:
What do you mean by previous implementation? The inners of ESPHome?

I was reviewing your code in HA and see that you translate ClimateMode to CURRENT_HVAC_*. If I were you I'd create another enum but is up to Otto approval.

OttoWinter replied:
Yeah, that would be better. HA also uses specialized enum/naming for action.

glmnet reviewed (commented):
in aioesphome api.proto you added this line

`ClimateMode action = 0;`

it does not match with the change in the api.proto here

Also I believe you cannot add 0, per proto buffers you should increment this value?

Also, are you making changes to esphome integration ha side? can you link your branch?

marcelveldt commented:
> in aioesphome api.proto you added this line
> 
> `ClimateMode action = 0;`
> 
> it does not match with the change in the api.proto here
> 
> Also I believe you cannot add 0, per proto buffers you should increment this value?
> 
> Also, are you making changes to esphome integration ha side? can you link your branch?

Fixed that typo. It's indeed 8 and not 0.

This is my proposal for the changes ha side:
https://github.com/marcelveldt/home-assistant/commit/e7088c59cfe4d6f4d6241a7cd62caccfed1521c4

OttoWinter commented on esphome/components/bang_bang/bang_bang_climate.cpp:
```diff
@@ -49,31 +49,31 @@ climate::ClimateTraits BangBangClimate::traits() {
   return traits;
 }
 void BangBangClimate::compute_state_() {
-  if (this->mode != climate::CLIMATE_MODE_AUTO) {
```
What are these changes in bang_bang controller for?

If I see correctly, the new changes would change the semantics of the bang_bang controller - only allowing heating when the device is in a specific mode? Please undo those changes and split them into a new PR and explain the change there.

marcelveldt replied:
Thanks for the review and changes! Offcourse I'm fine with your adjustments 👍 
About the change in bang_bang controller. There was a bug in there, which I tried to fix: 

If the mode is not auto (so heating or cooling) it just ignores the boundaries setup so it just keeps on cooling or heating, regardless of the temperature ranges.

OttoWinter replied:
That is how it's supposed to work - if you set it to heat or cool that means you want to manually override the internal logic and force it in that mode.

marcelveldt replied:
aha! Now I understand. I was in the understanding that, if set to auto, it would decide if it should cool OR heat. If set to heat, it will only allow heat and still account for the set points.

I've played a bit with some other thermostat implementations in Hass and they seem to work that way. But AUTO usually means that there's a schedule active.

The way it's implemented in Esphome is that setting it to heat will actually force the heating. Maybe that's something for the docs to clear up.

Feel free to remove my change in the bang_bang controller

OttoWinter committed:
Updates

- Use dedicated enum for action (otherwise it gets confusing because "auto" is not a valid action)
- Add field to tell HA that action is supported
- Revert semantic changes in bang_bang

OttoWinter reviewed (commented):
Hi, sorry for the long review delay first and foremost!

I had a look at the PR, and it looks good to me - I did some cleanup work making action a dedicated enum (otherwise it gets confusing because "auto" is not a valid action; believe me I have spent enough time working with the old HA climate model and that was one of it's biggest flaws: aliases)

I've also pushed changes here and to the aioesphomeapi PR here: https://github.com/esphome/aioesphomeapi/pull/2

If you're fine with these changes I'm glad to merge them

glmnet commented:
Great, I hope this one gets into dev soon, I was working on https://github.com/esphome/feature-requests/issues/333 but I'm waiting for this one first

OttoWinter committed:
Conditional print
